File: src/ios/com/nexj/bluetooth/parser/PacketParser.m
//
//  PacketParser.m
//  DevicePlugin
//
//  Created by Cait Potter on 1/14/13.
//  Copyright (c) 2013 __MyCompanyName__. All rights reserved.
//

#import "PacketParser.h"
#import "NJLog.h"

@implementation PacketParser

- (id)init
{
    if( self = [ super init ] )
    {
        return self;
    }
    return nil;
}

- (id)dataToJSON:(NSInputStream *)inputStream :(NSOutputStream *)outputStream
{
    // There is probably a better way to do this.
    if([self conformsToProtocol:@protocol(PacketParserProtocol)])
    {
        [ NSException raise:NSInvalidArgumentException
                      format:@"PacketParser does not conform to "
                             @"PacketParserProtocol" ];
    }
    else
    {
        id header;
        id packet;
        id json;

        // Header
        [ self preHeader:inputStream :outputStream ];
        @try
        {
            header = [ self createHeader:inputStream ];
        }
        @catch (NSException *e)
        {
            NJLog( self, @"Error occured during header creation: %@",
                   e.reason );
            @throw e;
        }
        [ self postHeader:inputStream :outputStream :header ];

        // Body
        [ self preBody:inputStream :outputStream :header ];
        @try
        {
            packet = [ self createPacket:header :inputStream ];
        }
        @catch (NSException *e)
        {
            NJLog( self, @"Error occured during packet creation: %@",
                   e.reason );
            @throw e;
        }
        [ self postBody:inputStream :outputStream :packet ];

        // Parsing
        [ self preParsing:inputStream :outputStream :packet ];
        @try
        {
            json = [ self parsePacket:packet ];
        }
        @catch (NSException *e)
        {
            NJLog( self, @"Error occurred during JSON object creation: %@",
                   e.reason );
            [ NSException raise:@"IOException"
                          format:e.reason ];
        }
        [ self postParsing:inputStream :outputStream :packet ];

        return json;
    }
}

- (void)copy:(NSOutputStream *)outputStream :(NSInputStream *)inputStream :(int)nLength
{
    int nCount;
    int b;
    uint8_t buffer[0x200];
    for( nCount = 0; nCount < nLength; ++nCount )
    {
        b = [ inputStream read:buffer maxLength:sizeof(buffer) ];
        if( b < 0 ) {
            // Read error occured. Throw?
            break;
        }
        if( [ outputStream write:buffer maxLength:b ] <= 0 ) {
            // Write error occured. Throw?
            break;
        }
    }
}

// Stub implementations (TODO: use NJError() to log these
- (void)respond:(id)header :(NSInputStream *)inputStream :(NSOutputStream *)outputStream
{
    NJError( self, @"method `-respond' not implemented"  );
}

- (id) parsePacket:(id)packet
{
    NJError( self, @"method `-parsePacket' not implemented" );
    return nil;
}

- (id)createPacket:(id)header :(NSInputStream *)is
{
    NJError( self, @"method `-createPacket' not implemented" );
    return nil;
}

- (id)createHeader:(NSInputStream *)is
{
    NJError( self, @"method `-createHeader' not implemented" );
    return nil;
}

- (void)preHeader:(NSInputStream *)is :(NSOutputStream *)os
{
}

- (void)postHeader:(NSInputStream *)is :(NSOutputStream *)os :(id)header
{
}

- (void)preBody:(NSInputStream *)is :(NSOutputStream *)os :(id)header
{
}

- (void)postBody:(NSInputStream *)is :(NSOutputStream *)os :(id)packet
{
}

- (void)preParsing:(NSInputStream *)is :(NSOutputStream *)os :(id)packet
{
}

- (void)postParsing:(NSInputStream *)is :(NSOutputStream *)os :(id)packet
{
}
@end

File: src/android/com/nexj/bluetooth/parser/PacketParser.java
package com.nexj.bluetooth.parser;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Map;

import org.json.JSONException;

import android.util.Log;

public abstract class PacketParser<P, H>
{
    protected final String m_sTag;

    public PacketParser(String sTag)
    {
        m_sTag = sTag;
    }

    public Map<String, Object> dataToPayload(InputStream is, OutputStream os, Map<String, Object> params) throws IOException
    {
        preHeader(is, os, (String)params.get("type"));

        H header = null;

        try
        {
            header = createHeader(is);
        }
        catch (IOException e)
        {
            Log.e(m_sTag, "Error occured during header creation", e);

            throw e;
        }

        postHeader(is, os, header);

        preBody(is, os, header);

        P packet = null;

        try
        {
            packet = createPacket(header, is);
        }
        catch (IOException e)
        {
            Log.e(m_sTag, "Error occured during packet creation", e);

            throw e;
        }

        postBody(is, os, packet);

        preParsing(is, os, packet);

        Map<String, Object> payload;

        try
        {
            payload = parsePacket(packet);
        }
        catch (JSONException e)
        {
            Log.e(m_sTag, "Error occured during JSON object creation", e);

            throw new IOException(e.getMessage());
        }

        postParsing(is, os, packet);

        return payload;
    }

    protected void preHeader(InputStream is, OutputStream os, String sDeviceType) throws IOException
    {
    }

    protected void postHeader(InputStream is, OutputStream os, H header) throws IOException
    {
    }

    protected void preBody(InputStream is, OutputStream os, H header) throws IOException
    {
    }

    protected void postBody(InputStream is, OutputStream os, P packet) throws IOException
    {
    }

    protected void preParsing(InputStream is, OutputStream os, P packet) throws IOException
    {
    }

    protected void postParsing(InputStream is, OutputStream os, P packet) throws IOException
    {
    }

    protected abstract Map<String, Object> parsePacket(P packet) throws JSONException;

    protected abstract P createPacket(H header, InputStream is) throws IOException;

    protected abstract H createHeader(InputStream is) throws IOException;

    public abstract String[] getSupportedDeviceTypes();

    // Copies the content of the inputstream into the outputstream
    protected void read(OutputStream os, InputStream is, int nLength) throws IOException
    {
        byte[] bData = new byte[1];

        for (int nCount = 0, nBytes = 0; nCount < nLength; nCount++)
        {
            nBytes = is.read(bData, 0, 1);

            if (nBytes == -1)
            {
                break;
            }

            os.write(bData, 0, 1);
        }
    }
}
